### DISCUSSION TITLE
Symfony Cache TagAwareAdapter problem with saving and then getting cacheItem

### DISCUSSION THREAD

**Discussion by ovgray** [symfony/symfony#49229] [Q&A] [answered]
The FilesystemTagAwareAdapter() lets me get, assign a value to, and save a cacheItem, and then retrieve the saved item with another get. But if I also tag the new item before saving, I cannot retrieve the saved item with another get.

When I run the following:
```
$this->cache = new \Symfony\Component\Cache\Adapter\FilesystemTagAwareAdapter();
$thingItem = $cache->getItem('thing');
$thingItem->set('thingValue');
// $thingItem->tag('msc');
$this->cache->save($thingItem);
$retrieved = $this->cache->getItem('thing')->get();
```
The value of $retrieved is "thingValue".

If I uncomment the "$thingItem->tag('msc');" line and run the code again, the value of $retrieved is null.

I would expect the value of "$retrieved" to be "thingValue" in both cases.

Can anyone explain the inconsistent behavior?

Edit:
I should say that this is on php 8.1 and windows 10.

The inconsistent behavior occurs in a phpunit TestCase test in a Symfony project I am working on. 

the inconsistent behavior does not occur in a function test using Symfony's KernelTestCase with 
`$this->cache = static::getContainer()->get('cache.app.taggable');`
So the fix for me would be to test the service in which I use the cache in a functional test. I still wonder why it does not work in the unit test, though.

**Comment by MatTheCat** [comment]
Could you provide a [reproducer](https://symfony.com/doc/current/contributing/code/reproducer.html)? I cannot replicate the issue by copy-pasting your code.

**Reply by ovgray** [reply]
Did you try to replicate in a windows environment? 

I also posted this question at [https://stackoverflow.com/questions/75341835/symfony-cache-tagawareadapter-problem-with-saving-and-then-getting-cacheitem](url). One of the comments I got said this:

`the tagging by default works via symlinks in an extra tag directory. if the OS doesn't allow symlinks, the cache entry will be removed:`

That comment pointed to [[github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Cache/…](https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Cache/Adapter/FilesystemTagAwareAdapter.php#L149-L150)](url)

I guess that explains why running the code under windows in a phpunit test does not work with the tagging line uncommented. But when I run it in a functional test as described, the code does work. So when symfony creates the cache.app.taggable service in a windows environment, it must do something to enable the tagging in a different way?

**Reply by MatTheCat** [reply]
> the tagging by default works via symlinks in an extra tag directory. if the OS doesn't allow symlinks, the cache entry will be removed

Nice I learnt something :ok_hand:

> when I run it in a functional test as described, the code does work.

Maybe your tests don’t use a filesystem adapter?

**Reply by ovgray** [reply]
See my answer below.

**Comment by ovgray** [comment, marked as answer]
Solved!

It appears that the cache.app.taggable that symfony creates in a windows environment is not a FilesystemTagAwareAdapter. It is a TagAwareAdapter wrapped around a TraceableAdapter wrapped around a FilesystemAdapter().

So if I create my cache this way:

`$this->cache = new TagAwareAdapter(new TraceableAdapter(new FilesystemAdapter());`

The code works as expected!

**Reply by MatTheCat** [reply]
Nice :+1: 

Be aware `FilesystemTagAwareAdapter` is supposed to be more efficient (cf. #30370) so it may be better to run your code inside an environment which supports symlink.
